src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -1,18 +1,654 @@
-        /* PLACEHOLDER METHOD */
+
+  private double targetHeadingDegrees;
+
+  double tagTargetHeading = 0.0;
+
+  // Create SwerveModules
+  public final SwerveModule frontLeft = new SwerveModule(
+      RobotMap.FRONT_LEFT_DRIVING_CAN_ID,
+      RobotMap.FRONT_LEFT_TURNING_CAN_ID,
+      DriveCal.FRONT_LEFT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  public final SwerveModule frontRight = new SwerveModule(
+      RobotMap.FRONT_RIGHT_DRIVING_CAN_ID,
+      RobotMap.FRONT_RIGHT_TURNING_CAN_ID,
+      DriveCal.FRONT_RIGHT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  public final SwerveModule rearLeft = new SwerveModule(
+      RobotMap.REAR_LEFT_DRIVING_CAN_ID,
+      RobotMap.REAR_LEFT_TURNING_CAN_ID,
+      DriveCal.BACK_LEFT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  public final SwerveModule rearRight = new SwerveModule(
+      RobotMap.REAR_RIGHT_DRIVING_CAN_ID,
+      RobotMap.REAR_RIGHT_TURNING_CAN_ID,
+      DriveCal.REAR_RIGHT_CHASSIS_ANGULAR_OFFSET_RAD);
+
+  private final Pigeon2 gyro = new Pigeon2(RobotMap.PIGEON_CAN_ID);
+  private ChassisSpeeds lastSetChassisSpeeds = new ChassisSpeeds(0.0, 0.0, 0.0);
+  public Optional<Pose2d> targetPose = Optional.empty();
+
+  public PoseBuffer poseBuffer = new PoseBuffer();
+
+  double KeepHeadingPID = 0.0;
+  double KeepHeadingFF = 0.0;
+
+  SwerveDriveOdometry odometry = new SwerveDriveOdometry(
+      DriveConstants.DRIVE_KINEMATICS,
+      Rotation2d.fromDegrees(Constants.PLACEHOLDER_DOUBLE),
+      getModulePositions());
+
+  public double throttleMultiplier = 1.0;
+
+  private double rotControllerInput = 0.0;
+
+  public MatchStateUtil matchState;
+
+  private InterpolatingDoubleTreeMap yawOffsetMap;
+
+  private InterpolatingDoubleTreeMap velocityToMultiplierMap;
+
+  public DriveSubsystem(MatchStateUtil matchState) {
+    intializeGyro();
+    this.matchState = matchState;
+
+    RobotConfig config;
+    try {
+      config = RobotConfig.fromGUISettings();
+      AutoBuilder.configure(
+        this::getPose, // Robot pose supplier
+        this::resetOdometry, // Method to reset odometry (will be called if your auto has a starting
+        // pose)
+        this::getCurrentChassisSpeeds, // ChassisSpeeds supplier. MUST BE ROBOT RELATIVE
+        (speeds, feedforwards) -> setOutputRobotRelativeSpeeds(speeds), // Method that will drive the robot given ROBOT RELATIVE
+        // ChassisSpeeds
+        new PPHolonomicDriveController(DriveCal.PATH_TRANSLATION_CONTROLLER, DriveCal.PATH_ROTATION_CONTROLLER),
+        config,
+        () -> {
+          // Boolean supplier that controls when the path will be mirrored for the red
+          // alliance
+          // This will flip the path being followed to the red side of the field.
+          // THE ORIGIN WILL REMAIN ON THE BLUE SIDE
+
+          var alliance = DriverStation.getAlliance();
+          if (alliance.isPresent()) {
+            return alliance.get() == DriverStation.Alliance.Red;
+          }
+          return false;
```

Comment:
i think this could be turned into a single boolean expression
return isPresent() && Red